### ISSUE TITLE
TFLite: Getting error "Expected bias tensor to be a vector" when trying to convert and quantize a model

### ISSUE THREAD

**Issue by ntreepoint** [tensorflow/tensorflow#43877]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): from pip
- TensorFlow version (use command below): 2.3.1
- Python version: 3.7.9
- Bazel version (if compiling from source): NA
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Describe the current behavior**
TensorFlow Lite fails to convert a FC4 model when quantization is activated.
I am getting this error when trying to perform full integer quantization with mixed 16 bits activations and 8 bits weights ( tf.lite.OpsSet.EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8) and when trying to use 8 bit quantization method (tf.lite.OpsSet.TFLITE_BUILTINS_INT8).)
I did several tests. First it is working if i convert a MobileNet V1 network, so the error depends on the network architecture.
By dichotomy, i have found the layer which causes the error. I guess the error is raised when there is a ADD (AddV2 in my case) operation right after a Conv2D operation. The converter try to merge Add operation as a bias in Conv2D. In my case, the tensor dimension of the constant data to be added is [1,1,1,64]. Maybe the converter is expecting [1,64] which could explain the error raised.

Please note that  8-bit quantization is working when using TensorFlow 2.1.0, so it looks like a regression introduced between 2.1.0 and 2.3.1.

**Describe the expected behavior**
I expect the converter to successfully convert the model with desired quantization.
 
**Standalone code to reproduce the issue**
Here is the code to reproduce the issue:
```
import tensorflow as tf
import tensorflow_datasets as tfds

import os

def representative_dataset_gen():
  ds = tfds.load("flic",shuffle_files=True, split='train')
  assert isinstance(ds, tf.data.Dataset)
  print(ds)
  num_calibration_steps = 10;
  for _ in range(num_calibration_steps):
    example = ds.take(1)
    for i in example:
        image = i["image"]
        name = i["moviename"]
        print(name)
        print("type: " + str(type(image)))
        image = tf.image.resize(image,size=[256,256])
        image = tf.expand_dims(image, axis=0)
        yield [image]


# Convert onnx model to tflite format

# convert it first in tensorflow frozen graph
if not os.path.exists("./convertedModels/tensorflow"):
    os.makedirs("./convertedModels/tensorflow")

#Then convert it in tflite format
converter = tf.compat.v1.lite.TFLiteConverter.from_frozen_graph('pretrained/model_frozen.pb', #TensorFlow freezegraph .pb model file
                                                      input_arrays=['FCN_1/AlexNet/mul'], # name of input arrays as defined in torch.onnx.export function before.
                                                      output_arrays=['FCN_1/Sum'],  # name of output arrays defined in torch.onnx.export function before.
                                                      input_shapes={ 'FCN_1/AlexNet/mul': [1,256,256,3]}
                                                      )

converter.optimizations = [tf.lite.Optimize.DEFAULT]

converter.target_spec.supported_ops = [tf.lite.OpsSet.EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8]

converter.inference_input_type = tf.int16  # or tf.uint8
converter.inference_output_type = tf.int16  # or tf.uint8


converter.representative_dataset = tf.lite.RepresentativeDataset(
    representative_dataset_gen)

tf_lite_model = converter.convert()
# save the converted model
open('fc4_quant.tflite', 'wb').write(tf_lite_model)

```
Here is the input model to convert:
[model_frozen.zip](https://github.com/tensorflow/tensorflow/files/5347801/model_frozen.zip)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem.
Here is the traces i got:

```
Traceback (most recent call last):
  File "convert_tflite.py", line 61, in <module>
    tf_lite_model = converter.convert()
  File "/home/arnaud/anaconda3/envs/FC4_CONV/lib/python3.7/site-packages/tensorflow/lite/python/lite.py", line 1970, in convert
    return super(TFLiteConverter, self).convert()
  File "/home/arnaud/anaconda3/envs/FC4_CONV/lib/python3.7/site-packages/tensorflow/lite/python/lite.py", line 1339, in convert
    result = self._calibrate_quantize_model(result, **flags)
  File "/home/arnaud/anaconda3/envs/FC4_CONV/lib/python3.7/site-packages/tensorflow/lite/python/lite.py", line 452, in _calibrate_quantize_model
    inference_output_type, allow_float, activations_type)
  File "/home/arnaud/anaconda3/envs/FC4_CONV/lib/python3.7/site-packages/tensorflow/lite/python/optimize/calibrator.py", line 98, in calibrate_and_quantize
    np.dtype(activations_type.as_numpy_dtype()).num)
RuntimeError: Expected bias tensor to be a vector.
```

**Comment by bhack** [comment]
Can you share your model definition?

**Comment by ntreepoint** [comment]
@bhack  model is available here: https://github.com/yuanming-hu/fc4

**Comment by bhack** [comment]
Can you extract a very minimal submodel to reproduce this? If not probably will require more bandwidth/time to investigate it on our side.

**Comment by ntreepoint** [comment]
@bhack  I am not familiar with the model but the interesting part is in build_model() method in squeeze_net.py.
For your information the problematic node is 'FCN_1/AlexNet/add'

**Comment by bhack** [comment]
Can you share a full standalone examplr that we can just copy and run or a Colab (with the Model export code)?

**Comment by ntreepoint** [comment]
@bhack sorry for that i am not familiar with the network nor Colab

**Comment by bhack** [comment]
I suppose that you have a minimal code to instantiate the net from the library and export It that It could simplu run and that you can share.

If not I suggest you to open a ticket in the net repository.

**Comment by ntreepoint** [comment]
@bhack, i have finally created a colab of the project. Here it is: https://colab.research.google.com/drive/14X0DxaKSXmNBTYJC1gXnQ_SJrr9S7jkc?usp=sharing

**Comment by bhack** [comment]
I don't see the convert code to reproduce your issue.

**Comment by ntreepoint** [comment]
@bhack. I thought providing only network inference was ok as i already provided code for conversion. I will add it on Colab.

**Comment by bhack** [comment]
Also I see you are using `tensorflow==1.15.0` it will be very hard to you to get TF lite bug support here for an old version.

**Comment by ntreepoint** [comment]
@bhack the model is compatible with TF 1.15 but the converter i use is from 2.3.1. I have updated the Colab with generation of frozen graph and conversion in TensorFlow Lite format.

**Comment by bhack** [comment]
In your colab I see: `AttributeError: module 'tensorflow' has no attribute 'contrib'`

**Comment by ntreepoint** [comment]
@bhack That's because you had to click on "RESTART RUNTIME" each time there is an upgrade of tensorflow via pip. I have changed the Colab so that the restart is done programmatically.

**Comment by bhack** [comment]
I had already restarted the runtime there was something else. I will re-run now

**Comment by Saduf2019** [comment]
@ntreepoint 
Is this still an issue.

**Comment by ntreepoint** [comment]
@Saduf2019 , yes, this is still an issue.

**Comment by Saduf2019** [comment]
@ymodak 
I ran the code and colab crashes, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/cf7e9fdf53892c37f30b64b8c551e9f7/untitled454.ipynb).

**Comment by talumbau** [comment]
It looks like the goal here is to run the colab to generate the saved model, and then run the TF Lite converter on that saved model to get the error. However, I find that I can't successfully run the colab here:

in order to generate the saved model. The colab is odd because it has a cell for installing tf-nightly but then another cell for installing tf 1.15. If I install tf-nightly, then restart the run time, I get the error listed [above](https://github.com/tensorflow/tensorflow/issues/43877#issuecomment-708392873) about no module `contrib`. If I execute the cell that pip installs tf version 1.15, I get the following error instead:

```
WARNING:tensorflow:
The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-d34842144e55> in <module>()
      8 import os
      9 from tensorflow.python.tools import freeze_graph
---> 10 import tensorflow_datasets as tfds

3 frames
/usr/local/lib/python3.6/dist-packages/tensorflow_datasets/core/tf_compat.py in ensure_tf_install()
     57   if tf_version < min_tf_version:
     58     raise ImportError(
---> 59         "This version of TensorFlow Datasets requires TensorFlow "
     60         f"version >= {MIN_TF_VERSION}; Detected an installation of version "
     61         f"{tf.__version__}. Please upgrade TensorFlow to proceed."

ImportError: This version of TensorFlow Datasets requires TensorFlow version >= 2.1.0; Detected an installation of version 1.15.0. Please upgrade TensorFlow to proceed.

---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
```

It's not clear to me how this colab is supposed to work. Are we generating the saved model with TF Nightly? Or with version 1.15?

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.
